File: tests/DesktopGrass.Win2D.Tests/SimTests/PropSpacingTests.cs
// PropSpacingTests.cs - §15.6 Minimum spacing between adjacent props.

using System;
using System.Collections.Generic;
using DesktopGrass.Win2D;
using Xunit;

namespace DesktopGrass.Win2D.Tests.SimTests;

public class PropSpacingTests
{
    private const double Monitor1920 = 1920.0;

    private readonly record struct Prop(double LeftEdge, double RightEdge);

    private static Sim BuildSim()
    {
        var sim = new Sim
        {
            Blades = Sim.GenerateBlades(Constants.CANONICAL_TEST_SEED, Monitor1920, Constants.DEFAULT_DENSITY),
            WindowHeight = Constants.STRIP_HEIGHT + Constants.HEADROOM,
            GroundY = Constants.STRIP_HEIGHT + Constants.HEADROOM,
        };
        sim.ResetAmbientGusts(Constants.CANONICAL_TEST_SEED, Monitor1920);
        sim.ResetEntities(Constants.CANONICAL_TEST_SEED);
        return sim;
    }

    private static double CactusHalfWidth(in Blade b)
        => (b.CactusType != 0) ? b.CactusWidth * 1.55 : b.CactusWidth * 0.5;

    private static double PineHalfWidth(in Blade b)
    {
        double hw = (b.TreeVariant == 1) ? b.PineWidth * 4.0 : b.PineWidth * 0.5;
        if (b.TreeBackground) hw *= Constants.TREE_BG_SCALE;
        return hw;
    }

    // Walk a kind+layer's props left-to-right and assert the right edge of
    // each is at least PROP_MIN_GAP_DIP behind the left edge of the next.
    // Generators emit props in BaseX order so a linear pass suffices.
    private static void RequireSpacing(IReadOnlyList<Prop> props, string label)
    {
        Assert.True(props.Count >= 1, $"{label}: expected at least one placement");
        for (int i = 1; i < props.Count; i++)
        {
            double gap = props[i].LeftEdge - props[i - 1].RightEdge;
            Assert.True(gap >= Constants.PROP_MIN_GAP_DIP,
                $"{label}: pair {i - 1}->{i} gap={gap:F3} < min {Constants.PROP_MIN_GAP_DIP:F3}");
        }
    }

    [Fact]
    public void DesertCactiKeepMinGapBetweenNeighbours()
    {
        var sim = BuildSim();
        sim.SetScene(Scene.Desert);

        var cacti = new List<Prop>();
        foreach (var b in sim.Blades)
        {
            if (!b.IsCactus) continue;
            double hw = CactusHalfWidth(b);
            cacti.Add(new Prop(b.BaseX - hw, b.BaseX + hw));
        }
        RequireSpacing(cacti, "cacti");
    }

    [Fact]
    public void WinterPinesKeepMinGapWithinSameLayer()
    {
        var sim = BuildSim();
        sim.SetScene(Scene.Winter);

        var fg = new List<Prop>();
        var bg = new List<Prop>();
        foreach (var b in sim.Blades)
        {
            if (!b.IsPine) continue;
            double hw = PineHalfWidth(b);
            var p = new Prop(b.BaseX - hw, b.BaseX + hw);
            if (b.TreeBackground) bg.Add(p);
            else                  fg.Add(p);
        }
        RequireSpacing(fg, "foreground pines");
        RequireSpacing(bg, "background pines");
    }

    [Fact]
    public void AutumnMaplesKeepMinGapBetweenNeighbours()
    {
        var sim = BuildSim();
        sim.SetScene(Scene.Autumn);

        var maples = new List<Prop>();
        foreach (var b in sim.Blades)
        {
            if (!b.IsMaple) continue;
            double hw = b.MapleCanopyRadius;
            maples.Add(new Prop(b.BaseX - hw, b.BaseX + hw));
        }
        RequireSpacing(maples, "maples");
    }

    [Fact]
    public void OceanCoralKeepMinGapBetweenNeighbours()
    {
        var sim = BuildSim();
        sim.SetScene(Scene.Ocean);

        var coral = new List<Prop>();
        foreach (var b in sim.Blades)
        {
            if (!b.IsCoral) continue;
            double hw = b.CoralWidth * 0.5;
            coral.Add(new Prop(b.BaseX - hw, b.BaseX + hw));
        }
        RequireSpacing(coral, "coral");
    }

    [Fact]
    public void SpacingRuleStillLeavesEachSceneWithAFewProps()
    {
        var sim = BuildSim();

        sim.SetScene(Scene.Desert);
        int cactusCount = 0;
        foreach (var b in sim.Blades) if (b.IsCactus) cactusCount++;
        Assert.True(cactusCount >= 1, $"Desert cactus count after gap rule: {cactusCount}");

        sim.SetScene(Scene.Winter);
        int pineCount = 0;
        foreach (var b in sim.Blades) if (b.IsPine) pineCount++;
        Assert.True(pineCount >= 3, $"Winter pine count after gap rule: {pineCount}");

        sim.SetScene(Scene.Autumn);
        int mapleCount = 0;
        foreach (var b in sim.Blades) if (b.IsMaple) mapleCount++;
        Assert.True(mapleCount >= 1, $"Autumn maple count after gap rule: {mapleCount}");

        sim.SetScene(Scene.Ocean);
        int coralCount = 0;
        foreach (var b in sim.Blades) if (b.IsCoral) coralCount++;
        Assert.True(coralCount >= 5, $"Ocean coral count after gap rule: {coralCount}");
    }
}

File: tests/DesktopGrass.Native.Tests/src/prop_spacing_tests.cpp
#include "TestHelpers.h"

#include "Sim.h"
#include "Constants.h"

#include <vector>

using namespace desktopgrass;

namespace {

constexpr double kMonitor1920 = 1920.0;
constexpr double kDensity     = 1.0;
constexpr uint64_t kSeed      = 0xDE5C70F6A55ED511ull;

struct Prop {
    double leftEdge;
    double rightEdge;
};

double cactus_half_width(const Blade& b) {
    return (b.cactusType != 0) ? b.cactusWidth * 1.55 : b.cactusWidth * 0.5;
}

double pine_half_width(const Blade& b) {
    double hw = (b.treeVariant == 1) ? b.pineWidth * 4.0 : b.pineWidth * 0.5;
    if (b.treeBackground) hw *= TREE_BG_SCALE;
    return hw;
}

// Walk the prop list left-to-right and verify that every adjacent pair has
// at least PROP_MIN_GAP_DIP between the right edge of one and the left edge
// of the next. The generators emit props in baseX order so a single linear
// pass is sufficient.
void require_spacing(const std::vector<Prop>& props, double minGap) {
    Assert::IsTrue(props.size() >= 1);
    for (std::size_t i = 1; i < props.size(); ++i) {
        const double gap = props[i].leftEdge - props[i - 1].rightEdge;
        Assert::IsTrue(gap >= minGap);
    }
}

}  // namespace

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace DesktopGrassNativeTests
{
TEST_CLASS(PropSpacingTests)
{
public:
TEST_METHOD(DesertCactiKeepAtLeastPROPMINGAPDIPBetweenNeighbours) {
    Sim sim = sim_init(kSeed, kMonitor1920, kDensity);
    sim_set_scene(sim, Scene::Desert);

    std::vector<Prop> cacti;
    for (const Blade& b : sim.blades) {
        if (!b.isCactus) continue;
        const double hw = cactus_half_width(b);
        cacti.push_back({b.baseX - hw, b.baseX + hw});
    }
    require_spacing(cacti, PROP_MIN_GAP_DIP);
}

TEST_METHOD(WinterPinesKeepAtLeastPROPMINGAPDIPBetweenSameLayerNeighbours) {
    Sim sim = sim_init(kSeed, kMonitor1920, kDensity);
    sim_set_scene(sim, Scene::Winter);

    std::vector<Prop> fgPines;
    std::vector<Prop> bgPines;
    for (const Blade& b : sim.blades) {
        if (!b.isPine) continue;
        const double hw = pine_half_width(b);
        Prop p{b.baseX - hw, b.baseX + hw};
        if (b.treeBackground) bgPines.push_back(p);
        else                  fgPines.push_back(p);
    }
    require_spacing(fgPines, PROP_MIN_GAP_DIP);
    require_spacing(bgPines, PROP_MIN_GAP_DIP);
}

TEST_METHOD(AutumnMaplesKeepAtLeastPROPMINGAPDIPBetweenNeighbours) {
    Sim sim = sim_init(kSeed, kMonitor1920, kDensity);
    sim_set_scene(sim, Scene::Autumn);

    std::vector<Prop> maples;
    for (const Blade& b : sim.blades) {
        if (!b.isMaple) continue;
        const double hw = b.mapleCanopyRadius;
        maples.push_back({b.baseX - hw, b.baseX + hw});
    }
    require_spacing(maples, PROP_MIN_GAP_DIP);
}

TEST_METHOD(OceanCoralKeepAtLeastPROPMINGAPDIPBetweenNeighbours) {
    Sim sim = sim_init(kSeed, kMonitor1920, kDensity);
    sim_set_scene(sim, Scene::Ocean);

    std::vector<Prop> coral;
    for (const Blade& b : sim.blades) {
        if (!b.isCoral) continue;
        const double hw = b.coralWidth * 0.5;
        coral.push_back({b.baseX - hw, b.baseX + hw});
    }
    require_spacing(coral, PROP_MIN_GAP_DIP);
}

TEST_METHOD(PropSpacingRuleReducesButDoesnTDecimateThePopulation) {
    // Sanity check that gap rejection isn't aggressive enough to break the
    // existing "near-spec probability" tests — each scene should still place
    // at least a handful of props on a 1920-DIP window with canonical seed.
    Sim sim = sim_init(kSeed, kMonitor1920, kDensity);

    sim_set_scene(sim, Scene::Desert);
    int cactusCount = 0;
    for (const Blade& b : sim.blades) if (b.isCactus) ++cactusCount;
    Assert::IsTrue(cactusCount >= 1);

    sim_set_scene(sim, Scene::Winter);
    int pineCount = 0;
    for (const Blade& b : sim.blades) if (b.isPine) ++pineCount;
    Assert::IsTrue(pineCount >= 3);

    sim_set_scene(sim, Scene::Autumn);
    int mapleCount = 0;
    for (const Blade& b : sim.blades) if (b.isMaple) ++mapleCount;
    Assert::IsTrue(mapleCount >= 1);

    sim_set_scene(sim, Scene::Ocean);
    int coralCount = 0;
    for (const Blade& b : sim.blades) if (b.isCoral) ++coralCount;
    Assert::IsTrue(coralCount >= 5);
}
};
}
